### ISSUE TITLE
Strange behavior of sigma_8 derivatives

### ISSUE THREAD

**Issue by EiffL** [LSSTDESC/DifferentiableHOS#6]
Thank you @dlanzieri  for running the simulation script and getting a few examples of Jacobians.

I had a closer look at the results, trying to compare them to the jax-cosmo analytic jacobian in the case of the power spectrum, and I made the following plot:
![image (22)](https://user-images.githubusercontent.com/861591/110938464-34110200-8334-11eb-8276-562a1d7b75f9.png)
This plot shows the "normalized jacobians" that I compute by dividing the jacobians by the theory power spectrum. And I show here the case of the derivative of the power spectrum with respect to \sigma_8 and \Omega_c. The solid lines are jax-cosmo analytic jacobians, the dashed lines are FlowPM derivatives.

A few quick observations:
- The Omega_c derivatives look very good :-) there is a bit of scatter but that's because we are taking the average over only 10 realizations.

- There appears to be something wrong with the \sigma_8 derivatives, and it's scale dependent, that's very interesting. I would expect this normalized derivative of the PS with respect to sigma_8 to be mostly flat, similar to what we get in jax-cosmo. But I would not expect the derivative to be negative in parts of the PS....  
 
I have a theory for what might be happening.... In FlowPM (and FastPM I guess), \sigma_8 is used to define the amplitude of the **linear** matter power spectrum. In jax-cosmo, and CCL, \sigma_8 is used to define the amplitude of the full **non-linear** matter power spectrum at z=0. 

So.... if you assume that sigma_8 is mostly sensitive to linear scales, it shouldn't matter, but that's probably not true. And it would indeed introduce a difference in the power spectrum, but we may confuse it with resolution effect, it would also introduce a difference in the Jacobian, and that difference would indeed be scale-dependent. 

Other possibilities by order of likeliness:
- it's some "interesting" resolution effect.
- it's a bug in FlowPM
- it's a bug in jax-cosmo

I propose to investigate this difference in where sigma_8 is applied using jax-cosmo, essentially we can use sigma_8 to normalize the linear ps before adding the non-linear correction, and see what happens on the resulting Jacobian.

**Comment by EiffL** [comment]
Hummm actually I was wrong :-| jax-cosmo also applies the sigma_8 normalization on the linear power spectrum, before the non-linear correction:
https://github.com/DifferentiableUniverseInitiative/jax_cosmo/blob/fdffbf12eb12bcc086f01c2f8d24a24d1d9cb8ff/jax_cosmo/power.py#L48

So..... that's not it....

**Comment by EiffL** [comment]
@dlanzieri you are right that it might be due to problems in the FlowPM ray tracing. I had started to write tests based on lenstools for that. I'll try to finish that and add them to the repo.

Maybe on your side you could try to see how this changes with the resolution of the simulation?
So, looking at the inpact of the following parameters:
- nc
- plane_size

**Comment by dlanzieri** [comment]
@EiffL yes, great idea, I'll do it!

**Comment by EiffL** [comment]
@dlanzieri can you document here your experiments and results?

**Comment by dlanzieri** [comment]
@EiffL I compared the results of 3 different sets of simulation (32,64,128) to the jax-cosmo analytic Jacobian in the case of the power spectrum. I made your plots, and I found  that even though the Omega_c derivatives look very good for a 32 particles simulation, the situation changes when I increase the particle number to 64. The situation seems to improve using 128 particles , but unfortunately with that resolution I managed to make only one simulation, so it's very noisy and I don't know if it can help us. 
Here the plots:
![psnorm_32](https://user-images.githubusercontent.com/72620117/111163558-7e48ec00-859d-11eb-8497-0c93bb094ffa.png)
![psnorm_64](https://user-images.githubusercontent.com/72620117/111163563-7f7a1900-859d-11eb-9a36-b52e009bc3d9.png)
![psnorm_128](https://user-images.githubusercontent.com/72620117/111163566-7f7a1900-859d-11eb-88a0-a9b970d10226.png)
What do you think?

**Comment by EiffL** [comment]
Interesting! Thanks for this. Did you change both `nc` and `plane_size` here? So for instance, the plot for 64 is uses a 64x64x640 simulation volume, and a 64x64 plane?

**Comment by dlanzieri** [comment]
Hi @EiffL I've added a new version of the notebook [jacobian_powerspectrum.ipynb](https://github.com/LSSTDESC/DifferentiableHOS/blob/31ab21140718eec489c4b3a445a0ef07d60e72d4/notebooks/jacobian_powerspectrum.ipynb) where I compared the Jacobian computed by using our FlowPM tool (modified by all the adjustments we made in the last few weeks) with the one computed using the Jax-cosmo library (where we defined recently a new redshift distribution for a single source). 
This is how the angular power spectra look like:
![momcomp](https://user-images.githubusercontent.com/72620117/118684028-27ac9800-b802-11eb-943d-8cbb3f81f552.png)
Then, to compute the Jacobian I selected a range of ell in the following way:
```ruby
@tf.function
def compute_jacobian(Omega_c, sigma8):
    """ Function that actually computes the Jacobian of a given statistics
    """
    params = tf.stack([Omega_c, sigma8])
    with tf.GradientTape() as tape:
        tape.watch(params)
        m, lensplanes, r_center, a_center = compute_kappa(params[0], params[1])
        ell, power_spectrum= DHOS.statistics.power_spectrum(m[0,:,:,-1],field_size,
                                                         field_npix)
        
        k1 =tf.where(ell < 2000 ,False, True)
        ell=tf.boolean_mask(ell, tf.math.logical_not(k1))
        power_spectrum =tf.boolean_mask(power_spectrum, tf.math.logical_not(k1))
        
    return m,lensplanes, r_center, a_center, tape.jacobian(power_spectrum, params,
                             experimental_use_pfor=False), ell, power_spectrum 
```
(This is how the angular power spectra look like in this selected range of ell):
![momcomp1](https://user-images.githubusercontent.com/72620117/118685176-2760cc80-b803-11eb-8f05-b563fba2bdc9.png)
and this is the normalized Jacobian (I averaged with only 3 realisations for now ) 
![momcomp2](https://user-images.githubusercontent.com/72620117/118686681-8246f380-b804-11eb-8999-beded3201657.png)
Even if there is still a mismatch between the two methods, the sigma_8 derivatives are not zero now (look at the previous plots). 
I have two doubts:
1. I'm not sure that I selected the ell range in the proper way 
2. If I did it correctly, looking at the two angular power spectra in the selected range of ell, can we aspect a better agreement between the Jacobians?

**Comment by EiffL** [comment]
Hi @dlanzieri , thanks for this! One comment, since you normalize the jacobian by the theory power spectrum, we expect that if the cls are biased so will be the normalized jacobian, right?

Can you plot the jacobian divided by their corresponding Cl? 
This way this should compensate effects that are just a multiplicative factor on the power spectrum.


Another point, instead of restricting the range of ell, you can also reduce the dimension of the Cell by "rebinning" the array, so for instance, imagine you have an array of size 32, if you use the average of 2 consecutive points, you can turn it into an array of size 16. Does that makes sense?
And you can do that with a higher rebinning factor too.

**Comment by dlanzieri** [comment]
@EiffL I didn't think of that! :O
![momcomp5](https://user-images.githubusercontent.com/72620117/118781806-bc100c80-b88d-11eb-9d8e-339df663e688.png)

I'll try changing the binning too!

**Comment by EiffL** [comment]
Aaaahhhh that's looking pretty good!

**Comment by EiffL** [comment]
@dlanzieri for the rebinning, checkout this stack overflow post: https://stackoverflow.com/questions/8090229/resize-with-averaging-or-rebin-a-numpy-2d-array
The first answer is how I usually do it

**Comment by dlanzieri** [comment]
Thank you @EiffL, it was very useful. But I was wondering,  If I just reduce the dimension of the Cell by "rebinning" the array, I have to consider also very large ell, with our setting of simulations, can we go to such small scales? are these results reliable? 
![momcomp6](https://user-images.githubusercontent.com/72620117/118841492-a66c0880-b8c8-11eb-817e-94660c2e0130.png)

**Comment by EiffL** [comment]
if you plot the ell axis in log, it will be easier to compare to  the Cell plot :-)

But I think it looks pretty good! 

So, last time, if you remember, you were finding some widly different behaviour when changing the configuration of the simulation, so could you try to run it maybe at higher resolution and check that things are consistent at least on large scales, and improve on small scales?

**Comment by EiffL** [comment]
@dlanzieri once you have cleaned them up a bit more, can you post here your latest jacobian plots using PGD?
